packages/eas-cli/src/credentials/ios/appstore/provisioningProfileAdhoc.ts

```diff
@@ -1,9 +1,219 @@
+
+  for (const udid of udids) {
+    if (!alreadyAddedUdids.includes(udid)) {
+      const device = await Device.createAsync({ name: 'iOS Device (added by Expo)', udid });
+      alreadyAdded.push(device);
+    }
+  }
+
+  return alreadyAdded;
+}
+
+async function findDistCert(serial_number: string, isEnterprise: boolean) {
+  const certs = await Certificate.getAsync({
+    query: {
+      filter: {
+        certificateType: CertificateType.IOS_DISTRIBUTION,
+      },
+    },
+  });
+
+  if (serial_number === '__last__') {
+    return certs[certs.length - 1];
+  }
+
+  return certs.find(c => c.attributes.serialNumber === serial_number) ?? null;
+}
+
+async function findProfileByBundleId(bundleId: string, certSerialNumber: string) {
+  let expoProfiles = await getProfilesForBundleId(bundleId);
+  expoProfiles = expoProfiles.filter(
+    profile => profile.attributes.profileType === ProfileType.IOS_APP_INHOUSE
+  );
+
+  expoProfiles = expoProfiles.filter(profile => {
+    return (
+      profile.attributes.name.startsWith('*[expo]') &&
+      profile.attributes.profileState !== ProfileState.EXPIRED
+    );
+  });
+
+  const expoProfilesWithCert: Profile[] = [];
+  // find profiles associated with our development cert
+  for (const profile of expoProfiles) {
+    const certificates = await profile.getCertificatesAsync();
+    if (certificates.some(cert => cert.attributes.serialNumber === certSerialNumber)) {
+      expoProfilesWithCert.push(profile);
+    }
+  }
+
+  if (expoProfilesWithCert) {
+    // there is an expo managed profile with our desired certificate
+    // return the profile that will be valid for the longest duration
+    return {
+      profile: expoProfilesWithCert.sort(sortByExpiration)[expoProfilesWithCert.length - 1],
+      didUpdate: false,
+    };
+  } else if (expoProfiles) {
+    // there is an expo managed profile, but it doesnt have our desired certificate
+    // append the certificate and update the profile
+    const isInHouse = (await Teams.getTeamInfoAsync()).type.toLowerCase() === 'in-house';
+    const distCert = await findDistCert(certSerialNumber, isInHouse);
+    if (!distCert) throw new Error('expected cert not found');
+    const profile = expoProfiles.sort(sortByExpiration)[expoProfiles.length - 1];
+    profile.attributes.certificates = [distCert];
+    return { profile: await profile.regenerateAsync(), didUpdate: true };
+  }
+
+  // there is no valid provisioning profile available
+  return { profile: null, didUpdate: false };
+}
+
+function sortByExpiration(a: Profile, b: Profile): number {
+  return (
+    new Date(a.attributes.expirationDate).getTime() -
+    new Date(b.attributes.expirationDate).getTime()
+  );
+}
+
+async function findProfileById(profileId: string, bundleId: string) {
+  let profiles = await getProfilesForBundleId(bundleId);
+  profiles = profiles.filter(
+    profile => profile.attributes.profileType === ProfileType.IOS_APP_ADHOC
+  );
```

Comment:
Please don't reassign variables, it's an antipattern. See my other comment re this.